homeassistant/components/izone/discovery.py

```diff
@@ -0,0 +1,119 @@
+"""Internal discovery service for  iZone AC."""
+
+import logging
+from asyncio import Event
+from typing import Dict
+
+import pizone
+
+from homeassistant.const import CONF_EXCLUDE, EVENT_HOMEASSISTANT_STOP
+from homeassistant.helpers import aiohttp_client
+from homeassistant.helpers.typing import ConfigType, HomeAssistantType
+from homeassistant.helpers.dispatcher import (
+    async_dispatcher_send, async_dispatcher_connect)
+
+from .const import (
+    DATA_CONFIG, DATA_DISCOVERY_SERVICE,
+    DISPATCH_CONTROLLER_DISCOVERED,
+    DISPATCH_CONTROLLER_DISCONNECTED, DISPATCH_CONTROLLER_RECONNECTED,
+    DISPATCH_CONTROLLER_UPDATE, DISPATCH_ZONE_UPDATE)
+
+_LOGGER = logging.getLogger(__name__)
+
+
+class DiscoveryService(pizone.Listener):
+    """Discovery data and interfacing with pizone library."""
+
+    def __init__(self, hass):
+        """Initialise discovery service."""
+        super().__init__()
+        self.hass = hass
+        self.controllers = {}  # type: Dict[str, pizone.Controller]
+        self.controller_ready = Event()
+
+        self.pi_disco = None
+
+        async def _controller_discovered(ctrl: pizone.Controller):
+            assert ctrl.device_uid not in self.controllers, \
+                "discovered device that already exists"
+            _LOGGER.debug("Controller discovered uid=%s", ctrl.device_uid)
+
+            conf = self.hass.data.get(DATA_CONFIG)  # type: ConfigType
+
+            # Filter out any entities excluded in the config file
+            if conf and ctrl.device_uid in conf[CONF_EXCLUDE]:
+                _LOGGER.info(
+                    "Controller UID=%s ignored as excluded",
+                    ctrl.device_uid)
+                return
+
+            self.controllers[ctrl.device_uid] = ctrl
+            self.controller_ready.set()
+
```

Comment:
We should signal the platform discover function here, right?

Reply:
I've stripped out all this code from here and put it into the climate device instead.

All that discovery does now is reinterprets the library's callbacks as signals.